hw3/02-charts.js

```diff
@@ -1,52 +1,133 @@
-  'rgba(75, 192, 192, 1)',
-  'rgba(153, 102, 255, 1)',
-  'rgba(255, 159, 64, 1)',
-  'rgba(159, 159, 159, 1)',
-  'rgba(83, 102, 255, 1)',
-  'rgba(40, 159, 64, 1)',
-  'rgba(210, 199, 199, 1)',
-  'rgba(78, 52, 199, 1)',
+  "rgba(54, 162, 235, 1)",
+  "rgba(255, 206, 86, 1)",
+  "rgba(255, 99, 132, 1)",
+  "rgba(75, 192, 192, 1)",
+  "rgba(153, 102, 255, 1)",
+  "rgba(255, 159, 64, 1)",
+  "rgba(159, 159, 159, 1)",
+  "rgba(83, 102, 255, 1)",
+  "rgba(40, 159, 64, 1)",
+  "rgba(210, 199, 199, 1)",
+  "rgba(78, 52, 199, 1)",
 ];
 
 // url for the Thrones API
-const url = 'https://thronesapi.com/api/v2/Characters';
+const url = "https://thronesapi.com/api/v2/Characters";
 
 const renderChart = () => {
-  const donutChart = document.querySelector('.donut-chart');
+  const donutChart = document.querySelector(".donut-chart");
 
   new Chart(donutChart, {
-    type: 'doughnut',
+    type: "doughnut",
     data: {
-      labels: ['label', 'label', 'label', 'label'],
+      labels: families,
       datasets: [
         {
-          label: 'My First Dataset',
-          data: [1, 12, 33, 5],
+          label: "Game of Thrones Houses",
+          data: numMembers,
           backgroundColor: backgroundColors,
           borderColor: borderColors,
           borderWidth: 1,
         },
       ],
     },
+    options: {
+      legend: {
+        display: false,
+      },
+    },
   });
 };
 
-renderChart();
+fetch(url)
+  .then((resp) => resp.json())
+  .then(findDistro)
+  .then(renderChart)
+  .catch((err) => console.log(`error getting data: ${err}`));
+
+let families = [];
+let numMembers = [];
+
+/* Sorts all the characters either into houses or non-house family group.
+ */
+function findDistro(data) {
+  // console.log(data);
+  for (let i = 0; i < data.length; i++) {
+    let family = data[i].family;
+    let index = matchHouse(family);
+
+    //consolidate families that don't have 'House ' in the name
+    if (index === -1 && !family.includes("House ")) {
+      family = "Non-House Family";
+      index = families.findIndex((e) => e.includes(family));
+    }
+
+    if (index < 0) {
+      families.push(family);
+      numMembers.push(1);
+    } else {
+      numMembers[index] += 1;
+    }
+  }
+  return data;
+}
+
+/*
+Looks for any houses that are already in the list, also check for any 
+house names that are similarly spelt
+*/
+function matchHouse(fam) {
+  //if the name is already in the list or its a part of another name
+  const i = families.findIndex((e) => e.includes(fam));
+  if (i > -1) return i;
+  //check if there are names spelt similarly (Typos)
+  const j = families.findIndex(
+    (e) =>
+      levenshteinDistance(
+        fam.replace("House ", ""),
+        e.replace("House ", "")
+      ) === 1
+  );
+  return j;
+}
+/*
+ Levenshtein Distance: Finds the number of differences between 2 strings
+ From https://www.tutorialspoint.com/levenshtein-distance-in-javascript
+*/
+function levenshteinDistance(str1 = "", str2 = "") {
```

Comment:
Yo, dawg. Above and beyond here. This seems like a smart way of detecting if the names of the chars are the same minus mispellings. Like targaryn vs targaryan etc. 
Major kudos here

Reply:
Thanks!